lib/msal-common/src/client/DeviceCodeClient.ts

```diff
@@ -0,0 +1,181 @@
+import {ServerAuthorizationTokenResponse} from "../server/ServerAuthorizationTokenResponse";
+
+/**
+ * OAuth2.0 Device code client
+ */
+export class DeviceCodeClient extends BaseClient {
+
+    private authority: Authority;
+
+    constructor(configuration: Configuration){
+        super(configuration);
+    }
+
+    /**
+     * Gets device code from device code endpoint, calls back to with device code response, and
+     * polls token endpoint to exchange device code for tokens
+     * @param request
+     */
+    public async acquireToken(request: DeviceCodeRequest): Promise<string> {
+        this.authority = await this.createAuthority(request.authority);
+        const deviceCodeResponse: NetworkResponse<DeviceCodeResponse> = await this.getDeviceCode(request);
+        request.deviceCodeCallback(deviceCodeResponse.body);
+        const response: ServerAuthorizationTokenResponse = await this.acquireTokenWithDeviceCode(
+            request,
+            deviceCodeResponse.body);
+
+        // TODO handle response
+        return JSON.stringify(response);
+    }
+
+    /**
+     * Creates device code request and executes http GET
+     * @param request
+     */
+    private async getDeviceCode(request: DeviceCodeRequest): Promise<NetworkResponse<DeviceCodeResponse>> {
+
+        const deviceCodeUrl = this.createDeviceCodeUrl(request);
+        const headers = this.createDefaultLibraryHeaders();
+
+        return this.executeGetRequestToDeviceCodeEndpoint(deviceCodeUrl, headers);
+    }
+
+    /**
+     * Executes GET request to device code endpoint
+     * @param deviceCodeUrl
+     * @param headers
+     */
+    private async executeGetRequestToDeviceCodeEndpoint(deviceCodeUrl: string, headers: Map<string, string>): Promise<NetworkResponse<DeviceCodeResponse>>{
+
+        return this.networkClient.sendGetRequestAsync<DeviceCodeResponse>(
+            deviceCodeUrl,
+            {
+                headers: headers
+            });
+    }
+
+    /**
+     * Create device code endpoint url
+     * @param request
+     */
+    private createDeviceCodeUrl(request: DeviceCodeRequest) : string {
+        const queryString: string = this.createQueryString(request);
+
+        // TODO add device code endpoint to authority class
+        return `${this.authority.canonicalAuthority}${Constants.DEVICE_CODE_ENDPOINT_PATH}?${queryString}`;
+    }
+
+    /**
+     * Create device code endpoint query parameters and returns string
+     */
+    private createQueryString(request: DeviceCodeRequest): string {
+
+        const parameterBuilder: RequestParameterBuilder = new RequestParameterBuilder();
+        parameterBuilder.addClientId(this.config.authOptions.clientId);
+
+        const scopes = RequestValidator.validateAndGenerateScopes(
+            request.scopes,
+            this.config.authOptions.clientId
+        );
+        parameterBuilder.addScopes(scopes);
+
+        return parameterBuilder.createQueryString();
+    }
```

Comment:
You could find an abstraction for this the same way you can for AuthorizationCodeClient.